### ISSUE TITLE
More topics control

### ISSUE THREAD

**Issue by MayhemYDG** [SOHU-Co/kafka-node#145]
Is it possible to add such functionally so that we can pause/resume specific topics consumption?
And is it possible to specify the topic a message comes from?

```js
consumer.pauseTopics(['topic1']);
consumer.resumeTopics(['topic2']);

consumer.on('message', function(message, topic) {
  // handle message differently depending on the topic
});
```

**Comment by haio** [comment]
The pause/resume topics maybe use addTopics/removeTopics instead.
The `message` is a object `{ topic: 'topic', value: 'sth', offset: 1, partition: 0, key: -1 }`, you can get the topic it comes from.

**Comment by MayhemYDG** [comment]
Okay, thanks.
The code suggests that `message` is either a string or a buffer though.
https://github.com/SOHU-Co/kafka-node/blob/master/lib/client.js#L109-L112

**Comment by haio** [comment]
Yeah, this this depends on your encoding option in consumer, the default is utf8.

**Comment by MayhemYDG** [comment]
Alright, I took some time to play around with kafka-node.

`removeTopics`/`addTopics` are not equal alternatives to `pause`/`resume`:
In my test I have a producer sending a counter every second.
The consumer logs the count in the console.
If I `pause`, then `resume` 5 seconds later, I get something like this:
```
1
2
3
(pause)
(resume)
4
5
6
...
```
but if I replace `pause`/`resume` with `removeTopics`/`addTopics`, I get something like that instead:
```
1
2
3
(removeTopics)
(addTopics)
1
2
3
4
5
6
...
```
I can decrease the `autoCommitIntervalMs` but I'll still get at least one message once again, which is not what I desire.

**Comment by MayhemYDG** [comment]
Here's the test case files.
producer.js:
```js
var kafka = require('kafka-node');
var client = new kafka.Client('localhost:2181');
var producer = new kafka.Producer(client);

client.once('ready', function() {
  var count = 0;
  var sendCount = function() {
    console.log(count);
    producer.send([{ topic: 'test', messages: '' + count++ }], function(err) {
      if (err) {
        console.error(arguments);
      }
    });
    setTimeout(sendCount, 1000);
  };
  sendCount();
});
```


consumer.js:
```js
var kafka = require('kafka-node');
var client = new kafka.Client('localhost:2181');

var topic = 'test';
var consumer = new kafka.Consumer(client, [], { groupId: 'consumer' });

consumer.on('message', function(msg) {
  console.log('message', msg.value);
});

consumer.addTopics([topic], function() {
  console.log('addTopics', arguments);
});

setTimeout(function() {
  consumer.removeTopics([topic], function() {
    console.log('removeTopics', arguments);
    setTimeout(function() {
      consumer.addTopics([topic], function() {
        console.log('addTopics', arguments);
      });
    }, 3000);
  });
}, 3000);

// setTimeout(function() {
//   consumer.pause();
//   console.log('pause');
//   setTimeout(function() {
//     consumer.resume();
//     console.log('resume');
//   }, 3000);
// }, 3000);

process.on('SIGINT', function() {
  console.log('');
  consumer.close(true, function() {
    console.log('closing', arguments);
  });
});
```

`removeTopics`/`addTopics` requires either low `autoCommitIntervalMs ` or manual call to `commit`.

**Comment by haio** [comment]
I add the `pauseTopics&resumeTopics` in this [branch](https://github.com/haio/kafka-node/tree/pause-resume-topics), it's only work in low level Consumer, can you test it? 
The api will be 
```
consumer.pauseTopics(['topic1', {topic: 'topic2', partition: 0 }]);
consumer.resumeTopics(['topic2']);
```
some example can found [here](https://github.com/haio/kafka-node/blob/pause-resume-topics/test/test.consumer.js#L337-L374)

**Comment by MayhemYDG** [comment]
Just tried. Pausing works.
Resuming doesn't work, new messages since I paused never get consumed.

**Comment by haio** [comment]
I guess this is related with the `fetchMaxWaitMs` in consumer, after pausing the topic,if you send new message to the topic while it is waiting the response from sever, the message also will be consumed.

**Comment by MayhemYDG** [comment]
Well pause/resume work fine, messages sent before and after resuming are consumed properly, I don't see why resumeTopics shouldn't work the same way.

**Comment by haio** [comment]
Resume and resumeTopics are different technically. They work well in my
test, how do you test it?
Mayhem <notifications@github.com>于2015年2月2日星期一写道：

**Comment by MayhemYDG** [comment]
I use the same test case as above, but I use this block to try pauseTopics/resumeTopics:

```js
setTimeout(function() {
  consumer.pauseTopics([topic]);
  console.log('pause');
  setTimeout(function() {
    consumer.resumeTopics([topic]);
    console.log('resume');
  }, 3000);
}, 3000);
```

**Comment by haio** [comment]
You can debug by checking consumer.payloads and consumer.pausedPayloads
after pauseTopics/resumeTopics are called, they only modify the request
payloads of consumer
Mayhem <notifications@github.com>于2015年2月2日星期一写道：

**Comment by haio** [comment]
@MayhemYDG I should find the bug, try [this](https://github.com/haio/kafka-node/tree/d74001e8760700da4480d83008891c9aff5eb3d8)

**Comment by MayhemYDG** [comment]
`resumeTopics` works great now, messages are properly consumed after resuming.
I found another bug: paused topics won't have their offsets committed when closing the consumer.
```js
consumer.pauseTopics([topic]);
consumer.close(true, function() {
    console.log('closed');
  });
```

**Comment by haio** [comment]
Add more fix, it should be work.

**Comment by MayhemYDG** [comment]
Good job :+1:

**Comment by haio** [comment]
does it work?

**Comment by MayhemYDG** [comment]
Yes.

**Comment by haio** [comment]
This is available in 0.2.18

**Comment by MayhemYDG** [comment]
Thanks!
